File: 2020/23/crabarray.cpp
#include <iostream>
#include <ranges>
#include <cstddef>
#include <cstdlib>
#include <stdexcept>
#include <string_view>
#include <utility>
#include <vector>

class crab_array {
  public:
    struct node {
        std::size_t cw;
        std::size_t ccw;

        operator<=>(const node& other) const noexcept = default;

        operator bool() const noexcept
        {
            return *this != missing_node;
        }
    };

    constexpr static node missing_node = { 0, 0, };

    template<class R>
    crab_array(R&& values, std::size_t max) noexcept
        : buf_(max + 1, missing_node), focus_(0)
    {
        std::size_t highest_seen = 0;
        auto it_end = end(values);
        std::size_t last = 0;
        for (auto it = begin(values); it != it_end; ++it) {
            std::size_t this_node = *it;
            buf_[this_node] = { 0, last };
            if (!focus_)
                focus_ = this_node;
            if (last)
                buf_[last].cw = this_node;
            if (this_node > highest_seen)
                highest_seen = this_node;
            last = this_node;
        }

        for (std::size_t i = highest_seen + 1; i < max; ++i) {
            buf_[i] = { 0, last };
            if (!focus_)
                focus_ = i;
            buf_[last].cw = i;
            last = i;
        }

        if (last) {
            buf_[last].cw = focus_;
            buf_[focus_].ccw = last;
        }
    }

    std::size_t max() const noexcept
    {
        return buf_.size() - 1;
    }

    std::size_t focus() const noexcept
    {
        return focus_;
    }

    void move_cw() noexcept
    {
        if (focus_)
            focus_ = buf_[focus_].cw;
    }

    std::size_t insert_cw(std::size_t dest, std::size_t value)
    {
        if (buf_[value])
            throw std::runtime_error("duplicate insert!");
        buf_[value].cw = buf_[dest].cw;
        buf_[dest].cw = value;
        buf_[value].ccw = dest;
        return value;
    }

    std::size_t remove_cw()
    {
        if (!focus_)
            throw std::logic_error("remove_cw from empty array");
        std::size_t ret = buf_[focus_].cw;
        buf_[focus_].cw = buf_[ret].cw;
        buf_[buf_[focus_].cw].ccw = focus_;
        buf_[ret] = missing_node;
        return ret;
    }

    std::vector<std::size_t> to_vector() const
    {
        std::vector<std::size_t> result;
        if (!focus_)
            return result;
        std::size_t cur = focus_;
        do {
            result.push_back(cur);
            cur = buf_[cur].cw;
        } while (cur != focus_);
        return result;
    }

  private:
    std::vector<node> buf_;
    std::size_t focus_;
};

void crab_game(crab_array& cups)
{
    auto first = cups.remove_cw();
    auto second = cups.remove_cw();
    auto third = cups.remove_cw();

    auto seek = cups.focus() - 1;
    if (seek < 1)
        seek = cups.max();
    while (seek == first || seek == second || seek == third) {
        --seek;
        if (seek < 1)
            seek = cups.max();
    }

    seek = cups.insert_cw(seek, first);
    seek = cups.insert_cw(seek, second);
    seek = cups.insert_cw(seek, third);

    cups.move_cw();
}

int main(int argc, char** argv)
{
    if (argc != 4) {
        std::cerr << "Usage: " << argv[0] << " nums max iters\n";
        return 1;
    }

    std::size_t max = std::strtoul(argv[2], nullptr, 10);
    std::size_t iters = std::strtoul(argv[3], nullptr, 10);

    crab_array cups(std::ranges::views::transform(std::string_view(argv[1]),
      [](char c) { return int(c - '0'); }), max);

    for (std::size_t i = 0; i < iters; ++i) {
        crab_game(cups);
    }

    auto v = cups.to_vector();
    for (const auto& c : v) {
        std::cout << c;
    }
    std::cout << '\n';

    return 0;
}

File: 2020/23/crabarray.py
import sys

from typing import List

def make_cuparray(initial: List[int], up_to: int) -> List[int]:
    result = [0] * (up_to + 1)
    last_cup = 0
    for cup in initial:
        result[last_cup] = cup
        last_cup = cup
    for cup in range(max(initial) + 1, up_to + 1):
        result[last_cup] = cup
        last_cup = cup
    result[last_cup] = result[0]
    return result

def remove_clockwise(cups: List[int], focus: int) -> int:
    val = cups[focus]
    cups[focus] = cups[cups[focus]]
    cups[val] = 0 # "missing indicator"
    return val

def insert_clockwise(cups: List[int], target: int, cup: int) -> int:
    if cups[cup] != 0:
        raise ValueError('duplicate insert!')
    cups[cup] = cups[target]
    cups[target] = cup
    return cup

def to_sequence(cups: List[int], focus: int) -> List[int]:
    result = list()
    next_cup = focus
    while True:
        result.append(next_cup)
        next_cup = cups[next_cup]
        if next_cup == focus:
            break
    return result

def main(argv: List[str]) -> int:
    if len(argv) != 4:
        print(f'Usage: {argv[0]} init-cups n-cups n-iters', file=sys.stderr)
        return 1

    init_cups = [int(x) for x in argv[1].split(',')]
    n_cups = int(argv[2])
    n_iters = int(argv[3])

    cups = make_cuparray(init_cups, n_cups)
    focus = init_cups[0]

    for _ in range(n_iters):
        first = remove_clockwise(cups, focus)
        second = remove_clockwise(cups, focus)
        third = remove_clockwise(cups, focus)

        target = focus - 1
        if target < 1:
            target = n_cups
        while target in (first, second, third):
            target -= 1
            if target < 1:
                target = n_cups

        target = insert_clockwise(cups, target, first)
        target = insert_clockwise(cups, target, second)
        target = insert_clockwise(cups, target, third)

        focus = cups[focus]

    # print(','.join(str(x) for x in to_sequence(cups, focus)))

    one_cw = cups[1]
    one_cw2 = cups[one_cw]
    print(f'{one_cw} x {one_cw2} = {one_cw * one_cw2}')

    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))
